invokeai/app/services/download/download_default.py

```diff
@@ -0,0 +1,394 @@
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> DownloadJob:
+        """Create a download job and return its ID."""
+        with self._lock:
+            id = self._next_job_id
+            self._next_job_id += 1
+            job = DownloadJob(
+                id=id,
+                source=source,
+                dest=dest,
+                priority=priority,
+                access_token=access_token,
+            )
+            job.set_callbacks(
+                on_start=on_start,
+                on_progress=on_progress,
+                on_complete=on_complete,
+                on_cancelled=on_cancelled,
+                on_error=on_error,
+            )
+            self._jobs[id] = job
+            self._queue.put(job)
+        return job
+
+    def join(self) -> None:
+        """Wait for all jobs to complete."""
+        self._queue.join()
+
+    def list_jobs(self) -> List[DownloadJob]:
+        """List all the jobs."""
+        return list(self._jobs.values())
+
+    def prune_jobs(self) -> None:
+        """Prune completed and errored queue items from the job list."""
+        with self._lock:
+            to_delete = set()
+            for job_id, job in self._jobs.items():
+                if self._in_terminal_state(job):
+                    to_delete.add(job_id)
+            for job_id in to_delete:
+                del self._jobs[job_id]
+
+    def id_to_job(self, id: int) -> DownloadJob:
+        """Translate a job ID into a DownloadJob object."""
+        try:
+            return self._jobs[id]
+        except KeyError as excp:
+            raise UnknownJobIDException("Unrecognized job") from excp
+
+    def cancel_job(self, job: DownloadJob) -> None:
+        """
+        Cancel the indicated job.
+
+        If it is running it will be stopped.
+        job.status will be set to DownloadJobStatus.CANCELLED
+        """
+        with self._lock:
+            job.cancel()
+
+    def cancel_all_jobs(self, preserve_partial: bool = False) -> None:
+        """Cancel all jobs (those not in enqueued, running or paused state)."""
+        for job in self._jobs.values():
+            if not self._in_terminal_state(job):
+                self.cancel_job(job)
+
+    def _in_terminal_state(self, job: DownloadJob) -> bool:
+        return job.status in [
+            DownloadJobStatus.COMPLETED,
+            DownloadJobStatus.CANCELLED,
+            DownloadJobStatus.ERROR,
+        ]
+
+    def _start_workers(self, max_workers: int) -> None:
+        """Start the requested number of worker threads."""
+        self.stop()  # stop any running threads and reset the worker pool set
+        for i in range(0, max_workers):  # noqa B007
+            worker = threading.Thread(target=self._download_next_item, daemon=True)
+            self._logger.debug(f"Download queue worker thread {worker.name} starting.")
+            worker.start()
+            self._worker_pool.add(worker)
```

Comment:
We should make sure and join all threads before starting new ones. With this implementation, there's a chance that some of the new threads are the ones that get killed instead of just the old ones.

Reply:
I've jettisoned the STOP job framework and replaced it with the suggested threading.Event solution. I also put in logic to handle what happens when `stop()` is called while there are still jobs in the queue. Briefly:
1. `stop()` sets a flag that prevents the `download()` method from accepting any more requests. It raises a `ServiceInactiveException`
2. The queue is cleared of all jobs that have not yet been accepted by worker threads.
3. If there are running jobs, the code logs a message that we are waiting for active downloads to complete and calls queue.join() to wait for all active jobs to complete.
4. The code sets the stop Event to signal the worker threads to exit.

Reply:
> We should make sure and join all threads before starting new ones. With this implementation, there's a chance that some of the new threads are the ones that get killed instead of just the old ones.

This is a mute issue, but it isn't actually a problem if some "new" threads are killed instead of the "old" ones. Since all the threads do the same thing, we only care if there are threads leaking.